apm-opentracing/src/main/java/co/elastic/apm/opentracing/ApmScopeManager.java

```diff
@@ -29,29 +29,42 @@
 
     @Override
     public ApmScope activate(@Nonnull Span span, boolean finishSpanOnClose) {
+        // prevents other threads from setting ApmSpan.dispatcher to null
         final ApmSpan apmSpan = (ApmSpan) span;
-        doActivate(apmSpan.getSpan());
+        synchronized (span) {
+            doActivate(apmSpan.getSpan(), apmSpan.context().getTraceContext());
+        }
         return new ApmScope(finishSpanOnClose, apmSpan);
     }
 
-    private void doActivate(@Nullable Object span) {
+    private void doActivate(@Nullable Object span, Object traceContext) {
         // implementation is injected at runtime via co.elastic.apm.opentracing.impl.ScopeManagerInstrumentation
     }
 
     @Override
     @Nullable
     public ApmScope active() {
         final Object span = getCurrentSpan();
-        if (span == null) {
-            return null;
-        } else {
+        if (span != null) {
             return new ApmScope(false, new ApmSpan(span));
+        } else {
+            final Object traceContext = getCurrentTraceContext();
+            if (traceContext != null) {
+                return new ApmScope(false, new ApmSpan(new TraceContextSpanContext(traceContext)));
+            }
         }
+        return null;
```

Comment:
This can be a problem- since you `synchronize` on the ApmSpan object, you must make sure there is a single object representing a span. This method can return a different ApmSpan object (through an ApmScope), which means that the synchronization will not only be wrong, it may also introduces risk for deadlocks.
Why not holing the ApmScope as a ThreadLocal? Isn't that the scope of an ApmScope?

Reply:
That would make a lot of things easier but it would not allow getting the active span in case the span has been activated via the internal API and we would essentially have two distinct scope managers which don't interact with each other.